File: templates/python/scatter_marginal_rug.py
"""scatter_marginal_rug: 散点 + 轴边缘 rug 短线."""
import sys, os; sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', '_utils', 'python'))
import numpy as np
import matplotlib.pyplot as plt
from theme import apply_theme
from palette import cycle

def make_figure(title='Scatter with rug'):
    apply_theme()
    rng = np.random.default_rng(0)
    x = rng.normal(0, 1, 120); y = 0.7*x + rng.normal(0, 0.6, 120)
    fig, ax = plt.subplots()
    ax.scatter(x, y, s=25, color=cycle(0), alpha=0.7, edgecolors='w', linewidth=0.4)
    for xi in x: ax.plot([xi, xi], [ax.get_ylim()[0], ax.get_ylim()[0]+0.05*(ax.get_ylim()[1]-ax.get_ylim()[0])], color=cycle(0), linewidth=0.6, transform=ax.get_xaxis_transform(), clip_on=False)
    for yi in y: ax.plot([0, 0.012], [yi, yi], color=cycle(0), linewidth=0.6, transform=ax.get_yaxis_transform(), clip_on=False)
    ax.set_xlabel('x'); ax.set_ylabel('y'); ax.set_title(title)
    ax.grid(True, linestyle=':', alpha=0.5)
    fig.tight_layout()
    return fig

if __name__ == '__main__':
    fig = make_figure(); fig.savefig(__file__.replace('.py', '.png'), dpi=150)

File: templates/matlab/scatter_marginal_rug.m
function fig = scatter_marginal_rug()
    addpath(fullfile(fileparts(mfilename('fullpath')),'..','..','_utils','matlab'));
    apply_theme(); rng(0); c = palette('cat',1);
    x = randn(120,1); y = 0.7*x + 0.6*randn(120,1);
    fig = figure;
    scatter(x, y, 25, c, 'filled', 'MarkerFaceAlpha', 0.7, 'MarkerEdgeColor','w'); hold on;
    yl = ylim; xl = xlim;
    for i = 1:numel(x)
        plot([x(i) x(i)], [yl(1) yl(1)+0.03*diff(yl)], 'Color', c, 'LineWidth', 0.6);
        plot([xl(1) xl(1)+0.012*diff(xl)], [y(i) y(i)], 'Color', c, 'LineWidth', 0.6);
    end
    xlabel('x'); ylabel('y'); title('Scatter with rug'); grid on;
end
